File: sdks/sdk-go/client_test.go
package openfireblocks

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestSign(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/sign" || r.Method != http.MethodPost {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		if got := r.Header.Get("Authorization"); got != "Bearer k1" {
			t.Errorf("auth header = %q", got)
		}
		_ = json.NewEncoder(w).Encode(SignResult{
			RequestID: "r1", SignedTx: "0xabc", TxHash: "0xhash",
			From: "0xfrom", Status: "signed",
		})
	}))
	defer srv.Close()

	c := New(srv.URL, "k1")
	res, err := c.Sign(context.Background(), SignRequest{
		ChainID: 11155111, To: "0xabc", GasLimit: 21000, GasPrice: "1", Nonce: 0,
	})
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	if res.RequestID != "r1" || res.Status != "signed" {
		t.Errorf("unexpected result %+v", res)
	}
}

func TestSign_APIError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusForbidden)
		_, _ = w.Write([]byte(`{"error":"policy denied"}`))
	}))
	defer srv.Close()

	c := New(srv.URL, "k1")
	_, err := c.Sign(context.Background(), SignRequest{ChainID: 1, To: "0x", GasLimit: 21000, GasPrice: "1"})
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T", err)
	}
	if apiErr.StatusCode != http.StatusForbidden {
		t.Errorf("status = %d", apiErr.StatusCode)
	}
}

func TestGetTransactionEscapesID(t *testing.T) {
	var seen string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		seen = r.URL.EscapedPath()
		_, _ = w.Write([]byte(`{}`))
	}))
	defer srv.Close()

	c := New(srv.URL, "k1")
	if _, err := c.GetTransaction(context.Background(), "a b"); err != nil {
		t.Fatalf("GetTransaction: %v", err)
	}
	if seen != "/transactions/a%20b" {
		t.Errorf("escaped path = %q", seen)
	}
}

File: sdks/sdk-js/test/client.test.mjs
// SDK tests using Node's built-in test runner against the compiled dist output
// with a stubbed fetch (no network).
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { OpenFireblocksClient, OpenFireblocksError } from '../dist/index.js';

function stubFetch(handler) {
  return async (url, init) => handler(url, init);
}

test('sign sends auth header and returns parsed body', async () => {
  let seenUrl, seenInit;
  const client = new OpenFireblocksClient({
    baseUrl: 'http://gw',
    apiKey: 'k1',
    fetch: stubFetch((url, init) => {
      seenUrl = url;
      seenInit = init;
      return {
        ok: true,
        status: 200,
        text: async () =>
          JSON.stringify({
            requestId: 'r1',
            signedTx: '0xabc',
            txHash: '0xhash',
            from: '0xfrom',
            status: 'signed',
            broadcasted: false,
          }),
      };
    }),
  });

  const res = await client.sign({
    chainId: 11155111,
    to: '0x742d35Cc6634C0532925a3b844Bc9e7595f42bE',
    gasLimit: 21000,
    gasPrice: '20000000000',
    nonce: 0,
  });

  assert.equal(seenUrl, 'http://gw/sign');
  assert.equal(seenInit.method, 'POST');
  assert.equal(seenInit.headers.Authorization, 'Bearer k1');
  assert.equal(res.requestId, 'r1');
  assert.equal(res.status, 'signed');
});

test('non-2xx throws OpenFireblocksError with body', async () => {
  const client = new OpenFireblocksClient({
    baseUrl: 'http://gw/',
    apiKey: 'k1',
    fetch: stubFetch(() => ({
      ok: false,
      status: 403,
      text: async () => JSON.stringify({ error: 'policy denied' }),
    })),
  });

  await assert.rejects(
    () =>
      client.sign({
        chainId: 1,
        to: '0xabc',
        gasLimit: 21000,
        gasPrice: '1',
        nonce: 0,
      }),
    (err) => {
      assert.ok(err instanceof OpenFireblocksError);
      assert.equal(err.status, 403);
      assert.deepEqual(err.body, { error: 'policy denied' });
      return true;
    },
  );
});

test('getTransaction encodes the id into the path', async () => {
  let seenUrl;
  const client = new OpenFireblocksClient({
    baseUrl: 'http://gw',
    apiKey: 'k1',
    fetch: stubFetch((url) => {
      seenUrl = url;
      return { ok: true, status: 200, text: async () => JSON.stringify({ request_id: 'a b' }) };
    }),
  });
  await client.getTransaction('a b');
  assert.equal(seenUrl, 'http://gw/transactions/a%20b');
});
